packages/framework/attributor/src/mixinAttributor.ts

```diff
@@ -0,0 +1,216 @@
+ * @sealed
+ * @alpha
+ */
+export interface IRuntimeAttributor extends IProvideRuntimeAttributor {
+    /**
+     * @throws - If no AttributionInfo exists for this key.
+     */
+    get(key: AttributionKey): AttributionInfo;
+
+    /**
+     * @returns - Whether any AttributionInfo exists for the provided key.
+     */
+    has(key: AttributionKey): boolean;
+}
+
+/**
+ * @returns an IRuntimeAttributor for usage with `mixinAttributor`. The attributor will only be populated
+ * @alpha
+ */
+export function createRuntimeAttributor(): IRuntimeAttributor {
+    return new RuntimeAttributor();
+}
+
+/**
+ * Mixes in logic to load and store runtime-based attribution functionality.
+ * 
+ * The `scope` passed to `load` should implement `IProvideRuntimeAttributor`.
+ * 
+ * Existing documents without stored attributors will not start storing attribution information: if an
+ * IRuntimeAttributor is passed via scope to load a document that never previously had attribution information,
+ * that attributor's `has` method will always return `false`.
+ * @param Base - base class, inherits from FluidAttributorRuntime
+ * @alpha
+ */
+export const mixinAttributor = (
+    Base: typeof ContainerRuntime = ContainerRuntime,
+) => class ContainerRuntimeWithAttributor extends Base {
+        public static async load(
+            context: IContainerContext,
+            registryEntries: NamedFluidDataStoreRegistryEntries,
+            requestHandler?: ((request: IRequest, runtime: IContainerRuntime) => Promise<IResponse>) | undefined,
+            runtimeOptions: IContainerRuntimeOptions | undefined = {},
+            containerScope: FluidObject | undefined = context.scope,
+            existing?: boolean | undefined,
+            ctor: typeof ContainerRuntime = ContainerRuntimeWithAttributor as unknown as typeof ContainerRuntime
+        ): Promise<ContainerRuntime> {
+            const runtimeAttributor = (containerScope as FluidObject<IProvideRuntimeAttributor> | undefined)?.IRuntimeAttributor;
+            if (!runtimeAttributor) {
+                throw new UsageError("ContainerRuntimeWithAttributor must be passed a scope implementing IProvideRuntimeAttributor");
+            }
+
+            const pendingRuntimeState = context.pendingLocalState as { baseSnapshot?: ISnapshotTree };
+            const baseSnapshot: ISnapshotTree | undefined = pendingRuntimeState?.baseSnapshot ?? context.baseSnapshot;
+
+            const { audience, deltaManager } = context;
+            assert(audience !== undefined, "Audience must exist when instantiating attribution-providing runtime");
+
+            const runtime = await Base.load(
+                context,
+                registryEntries,
+                requestHandler,
+                runtimeOptions,
+                containerScope,
+                existing,
+                ctor
+            ) as ContainerRuntimeWithAttributor;
+            runtime.runtimeAttributor = runtimeAttributor as RuntimeAttributor;
```

Comment:
we could add runtime validation to this cast with a symbol private to this file if we wanted. It's not compile-time checked right now; only theoretically valid since `IRuntimeAttributor` is marked `@sealed`. Thoughts?